File: npm/src/utils/json.ts
import { Base } from './base';
import * as file from './file';
import Keywords from './keywords';
import { regexRefinery } from './refinery';

/**
 * Checks for `JSON` content
 * @param content String type content
 * @returns `true` if content is JSON type
 */
export function isContentJson(content: string, isFilePath?: boolean) {
    if(isFilePath)
        content = file.read(content).toString();
    try {
        JSON.parse(content);
        return true;
    }
    catch {
        return false;
    }
}
/**
 * Split by comma
 * @param content String type content
 * @returns `string[]`
 */
export function separateByComma(content: string) {
    return content.split(",");
}
/**
 * `JSON` specific functions
 */
export class Json extends Base {
    private jsonKeys: string[] = [];
    private jsonValues: string[] = [];
    private filePath: string | undefined;

    constructor(obj: any | string, isFilePath?: boolean) {
        super();
        if(typeof obj === "string" && isFilePath) {
            this.filePath = obj;
            this.obj = JSON.parse(file.read(this.filePath));
        }
        else if(typeof obj === "string" && !isFilePath) {
            this.obj = JSON.parse(obj)
            this.filePath = undefined;
        }
        else if(obj !== undefined) {
            this.obj = obj;
            this.filePath = undefined;
        }
    }
    /**
     * Allowed data types
     */
    private dataTypes = [
        "string",
        "number",
        "boolean"
    ]
    /**
     * Pushes valid JSON path keys
     * @param eachKey Key to be appended
     * @param prevKey Previous key
     */
    private pushKey(eachKey: string, prevKey: string) {
        if(prevKey === "")
            this.jsonKeys.push(eachKey);
        else
            this.jsonKeys.push(prevKey + "." + eachKey)
    }
    /**
     * Reads keys in recursive manner
     * @param jsonData `JSON` data
     * @param prevKey Previous key
     */
    private getKeys(jsonData: any, prevKey: string = "") {
        if(isContentJson(JSON.stringify(jsonData))) {
            Object.keys(jsonData).forEach(eachKey => {
                if(Array.isArray(jsonData[eachKey])) {
                    var allRaw: boolean = true;
                    for(let i = 0; i < jsonData[eachKey].length; i ++) {
                        if(this.dataTypes.includes(typeof jsonData[eachKey][i])) 
                            continue;
                        else {
                            allRaw = false
                            if(prevKey !== "")
                                this.getKeys(jsonData[eachKey][i], prevKey + "." + eachKey + "[" + i + "]");
                            else
                                this.getKeys(jsonData[eachKey][i], eachKey + "[" + i + "]");
                        }
                    }
                    if(allRaw && jsonData[eachKey].length !== 0)
                        for(let i = 0; i < jsonData[eachKey].length; i ++)
                        this.pushKey(eachKey + "[" + i + "]", prevKey);
                }
                else if(this.dataTypes.includes(typeof jsonData[eachKey]))
                    this.pushKey(eachKey, prevKey)
                else if(isContentJson(JSON.stringify(jsonData[eachKey]))) {
                    if(prevKey !== "")
                        this.getKeys(jsonData[eachKey], prevKey + "." + eachKey)
                    else
                        this.getKeys(jsonData[eachKey], eachKey)
                }
                else if(jsonData[eachKey] === null)
                    this.pushKey(eachKey, prevKey);
            })
        }
        else if(Array.isArray(jsonData)) {
            jsonData.forEach(eachElem => this.getKeys(eachElem))
        }
    }
    /**
     * Parses value from JSON 
     * @param key
     * @returns Value fetched from key
     */
    private getValueFromKey(key: string) {
        let value = this.obj;
        if(key.includes(".")) {
            let keyList: string[] = key.split(".");
            for(let j = 0; j < keyList.length; j ++) {
                if(keyList[j].includes("[") && keyList[j].includes("]")) {
                    let key: string = keyList[j].split("[")[0];
                    let index: number = parseInt(keyList[j].split("[")[1].split("]")[0]);
                    value = value[key][index];
                }
                else
                    value = value[keyList[j]];
            }
        }
        else if(key === "")
            value = this.obj;
        else
            value = value[key];
        return value;
    }
    /**
     * Parses value from JSON 
     * @param key 
     * @returns Value fetched from key
     */
    public parse(key?: string) {
        if(key?.startsWith("$."))
            key = key.split("$.")[1];
        if(key !== undefined)
            return this.getValueFromKey(key);
        else
            return this.obj;
    }
    /**
     * Reads all possible keys.
     * @returns Keys `string[]`
     */
    public getAllKeys() {
        this.getKeys(this.obj);
        return this.jsonKeys;
    }
    /**
     * Fetches all values with given `JSON` keys
     * @returns Values in `string[]`
     */
    public getAllValues() {
        if(this.jsonKeys.length === 0)
            this.getAllKeys();

        for(let i = 0; i < this.jsonKeys.length; i ++) {
            let value = this.getValueFromKey(this.jsonKeys[i]);
            this.jsonValues.push(value);
        }
        return this.jsonValues;
    }
    private removeWithSucComma(key: string, value: any, content: string) {
        var uniqueKeys = content.match(Keywords.removeWithSucComa(key, regexRefinery(value)))?.filter((value, index, array) => { return array.indexOf(value) === index });
        
        if(uniqueKeys !== undefined) {
            for(let i = 0; i < uniqueKeys?.length; i ++) {
                var val = regexRefinery(uniqueKeys[i]);
                content = content.replace(new RegExp(val, "g"), "");
            }
        }
        return content;
    }
    private removeWithPreComma(key: string, value: string, content: string) {
        var uniqueKeys = content.match(Keywords.removeWithPreComa(key, regexRefinery(value)))?.filter((value, index, array) => { return array.indexOf(value) === index });
        
        if(uniqueKeys !== undefined) {
            for(let i = 0; i < uniqueKeys?.length; i ++) {
                var val = regexRefinery(uniqueKeys[i]);
                content = content.replace(new RegExp(val, "g"), "");
            }
        }
        return content;
    }
    private removeRecursively(key: string, obj: any) {
        if(key.split(".").length === 1) {
            let stringObj: string = JSON.stringify(obj);
            let con = this.removeWithSucComma(key, obj[key], stringObj);
            if(! isContentJson(con)) {
                con = this.removeWithPreComma(key, obj[key], stringObj);
                con = this.removeWithSucComma(key, obj[key], con);
            }
            return con;
        }
        else {
            let curKey: string = key.split(".")[0];
            var a = this.removeRecursively(key.replace(curKey + ".", ""), obj[curKey]);
            
            if(a !== undefined)
                obj[curKey] = JSON.parse(a);
            else
                obj[curKey] = a;
            return JSON.stringify(obj);
        }
    }
    /**
     * Removes a key:value from the JSON context. Key will be JPath in `$.full.path` format
     * 
     * The function automatically deserializes before removing. So, no need to explicitely deserialize it.
     * @param key JPath to the key
     * @param content JSON `string` content from which to be removed.
     * @returns Resultant content in `JSON` object
     */
    public removeWithKey(key: string, content: string) {
        if(key.startsWith(Keywords.relativeJPath))
            key = key.replace(Keywords.relativeJPath, "");
        var value = this.parse(key);
        
        if(typeof value !== "object")
            this.obj = JSON.parse(this.removeRecursively(key, JSON.parse(content)));
        else {
            // Replacing with null first and then removing the key:value set
            var nulledObj: any = this.replaceRecursively(key, null, JSON.parse(content));
            this.obj = JSON.parse(this.removeRecursively(key, nulledObj));
        }
        return this.obj;
    }
    /**
     * Replace a JPath to a specified value. The function context is of JSON and cannot be used in CJSON context.
     * 
     * In order to use this in CJSON context, follow below steps:
     * <ol>
     * <li>Create CJson object</li>
     * <li>Deserialize</li>
     * <li>Deserialize</li>
     * <li>cjson.json?.replace("$.jpath", value, object)</li>
     * </ol>
     * @param key 
     * @param value 
     * @param jsonObject 
     * @returns 
     */
    public replace = (jPath: string, value: any, obj: any) => {
        this.obj = this.replaceRecursively(jPath, value, obj);
        return this.obj;
    }
    /**
     * Recursive function for replacing data in provided key.
     * @private
     * @param key 
     * @param obj 
     * @returns 
     */
    private replaceRecursively(key: string, value: any, obj: any): any {
        if(key.split(".").length === 1)
            obj[key] = value;
        else {
            let curKey: string = key.split(".")[0];
            obj[curKey] = this.replaceRecursively(key.replace(curKey + ".", ""), value, obj[curKey]);
        }
        return obj;
    }
}

File: python/cjson/src/utils/_json.py
from utils import file
from utils.keywords import Keywords
import json
from typing import Union
from types import NoneType
from utils._base import Base
import re

def is_content_json(content: str, is_file_path: bool = False):
    if is_file_path:
        content = file.read(content)
    
    try:
        json.loads(content)
        return True

    except:
        return False

class Json(Base):
    __json_keys: list[str] = []
    __json_values: list[str] = []
    __file_path: Union[str , None]
    __data_types: list[type] = [ str, int, bool ]

    def __init__(self, obj: Union[any , str], is_file_path: bool = False):
        if type(obj) is str and is_file_path:
            self.__file_path = obj
            file = open(self.__file_path)
            self._obj = json.load(file)
        elif type(obj) == str and not is_file_path:
            self._obj = json.loads(obj)
        else:
            self._obj = obj
            self.__file_path = None
    
    def parse(self, key: Union[ str, None ] = None ):
        ''' Run any `JPath` query in parsed `CJSON`/ `JSON`.

            Returns full `JSON` if `key` is not provided
        '''
        if key is not None and key.startswith("$."):
            key = key.replace("$.", "", 1)

        if key is None:
            return self._obj
        else:
            return self.__get_value_from_key(key)

    def get_all_keys(self):
        ''' Returns all possible `JSON` keys in parsed `JSON`
        '''
        self.__json_keys = []
        self.__get_keys(json_data=self._obj)
        return self.__json_keys

    def get_all_values(self):
        ''' Returns all possible `JSON` values in parsed `JSON`
        '''
        self.__json_values = []
        if len(self.__json_keys) == 0:
            self.get_all_keys()
        
        for i in range(0, len(self.__json_keys)):
            value: str = self.__get_value_from_key(self.__json_keys[i])
            self.__json_values.append(value)
        
        return self.__json_values

    def __push_key(self, cur_key: str, prev_key: str):
        if(prev_key == ""):
            self.__json_keys.append(cur_key)
        else:
            self.__json_keys.append(prev_key + "." + cur_key)

    def __get_keys(self, json_data: any, prev_key: str = ""):
        if(type(json_data) == dict):
            for each_key in json_data.keys():
                if(type(json_data[each_key]) in self.__data_types):
                    self.__push_key(each_key, prev_key)

                elif(type(json_data[each_key]) == dict):
                    if(prev_key != ""):
                        self.__get_keys(json_data[each_key], prev_key + "." + each_key)
                    else:
                        self.__get_keys(json_data[each_key], each_key)

                elif(type(json_data[each_key]) == list):
                    all_raw: bool = True

                    for each_in_elif in json_data[each_key]:
                        index = str(json_data[each_key].index(each_in_elif))

                        if(type(each_in_elif) in self.__data_types):
                            continue
                        else:
                            all_raw = False
                            
                            if(prev_key != ""):
                                self.__get_keys(each_in_elif, prev_key=prev_key + "." + each_key + "[" + index + "]")
                            else:
                                self.__get_keys(each_in_elif, prev_key=each_key + "[" + index + "]")
                    if(all_raw and len(json_data[each_key]) != 0):
                        self.__push_key(each_key, prev_key)

                elif(type(json_data[each_key]) == NoneType):
                    self.__push_key(each_key, prev_key)

        elif(type(json_data) == list):
            for each_in_elif in json_data:
                self.__get_keys(each_in_elif)
    
    def __get_value_from_key(self, key: str):
        ''' Returns all possible json keys in `CJSON` object
        '''
        value = self._obj
        if "." in key:
            key_list: list[str] = key.split(".")
            for i in range(0, len(key_list)):
                if "[" in key_list[i] and "]" in key_list[i]:
                    key_name: str = key_list[i].split("[")[0]
                    index: int = int(key_list[i].split("[")[1].split("]")[0])
                    value = value[key_name][index]
                else:
                    value = value[key_list[i]]
        else:
            value = value[key]
        return value

    def __remove_with_pre_comma(self, key: str, value: any, content: str):
        items: list[str] = re.findall(Keywords.remove_with_pre_coma(key=key, value=self._regex_refinery(str(value))), content)
        
        if len(items) != 0:
            items = list(set(items))
            for each_item in items:
                content = content.replace(each_item, "")
        return content
    
    def __remove_with_suc_comma(self, key: str, value: any, content: str):
        items: list[str] = re.findall(Keywords.remove_with_suc_coma(key=key, value=self._regex_refinery(str(value))), content)
        
        if len(items) != 0:
            items = list(set(items))
            for each_item in items:
                content = content.replace(each_item, "")
        return content
    
    def __remove_recursively(self, key: str, obj: any) -> str:
        if len(key.split(".")) == 1:
            string_obj: str = json.dumps(obj)
            # None value to be replaced with null
            if obj[key] is None:
                con = self.__remove_with_suc_comma(key, "null", string_obj)

                if not is_content_json(content=con):
                    con = self.__remove_with_pre_comma(key, "null", string_obj)
                    con = self.__remove_with_suc_comma(key, "null", con)

            else:
                con = self.__remove_with_suc_comma(key, obj[key], string_obj)

                if not is_content_json(content=con):
                    con = self.__remove_with_pre_comma(key, obj[key], string_obj)
                    con = self.__remove_with_suc_comma(key, obj[key], con)
            
            return con
        else:
            cur_key: str = key.split(".")[0]
            a: str = self.__remove_recursively(key.replace(cur_key + ".", ""), obj[cur_key])

            if a is not None:
                obj[cur_key] = json.loads(a)
            else:
                obj[cur_key] = a
            return json.dumps(obj=obj)
    
    def __replace_recursively(self, key: str, value: any, obj: any):
        if len(key.split(".")) == 1:
            obj[key] = value
        else:
            cur_key: str = key.split(".")[0]
            obj[cur_key] = self.__replace_recursively(key.replace(cur_key + ".", "", 1), value=value, obj=obj[cur_key])
            
        return obj
    
    def _remove_with_key(self, key: str, content: str):
        if(key.startswith(Keywords.relative_jpath)):
            key = key.replace(Keywords.relative_jpath, "")
        
        value = self.parse(key=key)

        if type(value) is not list or type(value) is not dict:
            self._obj = json.loads(self.__remove_recursively(key=key, obj=json.loads(content)))
        else:
            nulled_obj: any = self.__replace_recursively(key=key, value=None, obj=json.loads(content))
            # Updating JSON object
            self._obj = json.loads(self.__remove_recursively(key=key, obj=nulled_obj))
        # Updating JSON string variable
        self._content = json.dumps(self._obj)
            
        return self
    
    def replace(self, jpath: str, value: str, obj: any):
        self._obj = self.__replace_recursively(key=jpath, value=value, obj=obj)
        # Updating JSON string variable
        self._content = json.dumps(self._obj)
        return self._obj